File: colorpicker/src/main/java/org/catrobat/paintroid/colorpicker/HSVColorPickerView.java
/*
 * Paintroid: An image manipulation application for Android.
 * Copyright (C) 2010-2015 The Catrobat Team
 * (<http://developer.catrobat.org/credits>)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.catrobat.paintroid.colorpicker;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.LinearGradient;
import android.graphics.Paint;
import android.graphics.Point;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffXfermode;
import android.graphics.RectF;
import android.graphics.Shader;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;

public class HSVColorPickerView extends View {

	private static final float BORDER_WIDTH_PX = 1;

	private float panelSpacing = 10f;
	private float paletteCircleTrackerRadius = 5f;
	private float rectangleTrackerOffset = 2f;
	private float huePanelWidth = 30f;
	private float density = 1f;

	private Paint satValPaint;
	private Paint satValTrackerPaint;
	private Paint huePaint;
	private Paint hueTrackerPaint;
	private Paint borderPaint;

	private Shader valShader;
	private Shader hueShader;

	private int alpha = 0xff;
	private float hue = 360f;
	private float sat = 0f;
	private float val = 0f;

	private int borderColor = 0xff6E6E6E;

	private OnColorChangedListener onColorChangedListener;

	private float drawingOffset;

	private RectF drawingRect;
	private RectF satValRect;
	private RectF hueRect;

	private Point startTouchPoint = null;

	public HSVColorPickerView(Context context) {
		super(context);
		init();
	}

	public HSVColorPickerView(Context context, AttributeSet attrs) {
		super(context, attrs);
		init();
	}

	public HSVColorPickerView(Context context, AttributeSet attrs, int defStyle) {
		super(context, attrs, defStyle);
		init();
	}

	private void init() {
		density = getContext().getResources().getDisplayMetrics().density;
		paletteCircleTrackerRadius *= density;
		rectangleTrackerOffset *= density;
		huePanelWidth *= density;
		panelSpacing *= density;
		drawingOffset = paletteCircleTrackerRadius;

		satValPaint = new Paint();
		satValTrackerPaint = new Paint();
		huePaint = new Paint();
		hueTrackerPaint = new Paint();
		borderPaint = new Paint();

		satValTrackerPaint.setStyle(Paint.Style.STROKE);
		satValTrackerPaint.setStrokeWidth(2f * density);
		satValTrackerPaint.setAntiAlias(true);

		int sliderTrackerColor = 0xfff5f5f5;
		hueTrackerPaint.setColor(sliderTrackerColor);
		hueTrackerPaint.setStyle(Paint.Style.STROKE);
		hueTrackerPaint.setStrokeWidth(2f * density);
		hueTrackerPaint.setAntiAlias(true);

		drawingRect = new RectF();
		satValRect = new RectF();
		hueRect = new RectF();
	}

	@Override
	protected void onDraw(Canvas canvas) {
		if (drawingRect.width() <= 0 || drawingRect.height() <= 0) {
			return;
		}
		drawSatValPanel(canvas);
		drawHuePanel(canvas);
	}

	private void drawSatValPanel(Canvas canvas) {
		final RectF rect = satValRect;

		borderPaint.setColor(borderColor);
		canvas.drawRect(drawingRect.left, drawingRect.top, rect.right
						+ BORDER_WIDTH_PX, rect.bottom + BORDER_WIDTH_PX,
				borderPaint);

		if (valShader == null) {
			valShader = new LinearGradient(rect.left, rect.top, rect.left,
					rect.bottom, 0xffffffff, 0xff000000, Shader.TileMode.CLAMP);
		}

		int rgb = Color.HSVToColor(new float[]{hue, 1f, 1f});

		Shader satShader = new LinearGradient(rect.left, rect.top, rect.right,
				rect.bottom, Color.WHITE, rgb, Shader.TileMode.CLAMP);

		satValPaint.setXfermode(null);
		satValPaint.setShader(valShader);
		canvas.drawRect(rect, satValPaint);
		satValPaint.setShader(satShader);
		satValPaint.setXfermode(new PorterDuffXfermode(PorterDuff.Mode.MULTIPLY));
		canvas.drawRect(rect, satValPaint);

		Point p = satValToPoint(sat, val);
		satValTrackerPaint.setColor(0xff000000);
		canvas.drawCircle(p.x, p.y, paletteCircleTrackerRadius - 1f
				* density, satValTrackerPaint);

		satValTrackerPaint.setColor(0xffdddddd);
		canvas.drawCircle(p.x, p.y, paletteCircleTrackerRadius,
				satValTrackerPaint);
	}

	private void drawHuePanel(Canvas canvas) {
		final RectF rect = hueRect;

		borderPaint.setColor(borderColor);
		canvas.drawRect(rect.left - BORDER_WIDTH_PX, rect.top
						- BORDER_WIDTH_PX, rect.right + BORDER_WIDTH_PX,
				rect.bottom + BORDER_WIDTH_PX, borderPaint);

		if (hueShader == null) {
			hueShader = new LinearGradient(rect.left, rect.top, rect.left,
					rect.bottom, buildHueColorArray(), null,
					Shader.TileMode.CLAMP);
			huePaint.setShader(hueShader);
		}

		canvas.drawRect(rect, huePaint);

		float rectHeight = 4 * density / 2;
		Point p = hueToPoint(hue);
		RectF r = new RectF();
		r.left = rect.left - rectangleTrackerOffset;
		r.right = rect.right + rectangleTrackerOffset;
		r.top = p.y - rectHeight;
		r.bottom = p.y + rectHeight;

		canvas.drawRoundRect(r, 2, 2, hueTrackerPaint);
	}

	@Override
	public boolean onTouchEvent(MotionEvent event) {
		boolean update = false;
		switch (event.getAction()) {
			case MotionEvent.ACTION_DOWN:
				startTouchPoint = new Point((int) event.getX(), (int) event.getY());
				update = moveTrackersIfNeeded(event);
				break;

			case MotionEvent.ACTION_MOVE:
				update = moveTrackersIfNeeded(event);
				break;

			case MotionEvent.ACTION_UP:
				startTouchPoint = null;
				update = moveTrackersIfNeeded(event);
				break;
		}

		if (update) {
			invalidate();
			onColorChanged();
			return true;
		}
		return super.onTouchEvent(event);
	}

	private boolean moveTrackersIfNeeded(MotionEvent event) {
		if (startTouchPoint == null) {
			return false;
		}
		boolean update = true;

		int startX = startTouchPoint.x;
		int startY = startTouchPoint.y;

		if (hueRect.contains(startX, startY)) {
			hue = pointToHue(event.getY());
		} else if (satValRect.contains(startX, startY)) {
			float[] result = pointToSatVal(event.getX(), event.getY());
			sat = result[0];
			val = result[1];
		} else {
			update = false;
		}

		return update;
	}

	@Override
	protected void onSizeChanged(int w, int h, int oldw, int oldh) {
		super.onSizeChanged(w, h, oldw, oldh);

		drawingRect = new RectF();
		drawingRect.left = drawingOffset + getPaddingLeft();
		drawingRect.right = w - drawingOffset - getPaddingRight();
		drawingRect.top = drawingOffset + getPaddingTop();
		drawingRect.bottom = h - drawingOffset - getPaddingBottom();

		setUpSatValRect();
		setUpHueRect();
	}

	@Override
	protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
		int size = MeasureSpec.getSize(widthMeasureSpec) - getPaddingStart() - getPaddingEnd();
		setMeasuredDimension(size, size);
	}

	private static float clamp(float val, float min, float max) {
		return Math.max(min, Math.min(max, val));
	}

	private Point satValToPoint(float sat, float val) {
		final RectF rect = satValRect;
		final float height = rect.height();
		final float width = rect.width();

		Point p = new Point();
		p.x = (int) (sat * width + rect.left);
		p.y = (int) ((1f - val) * height + rect.top);

		return p;
	}

	private Point hueToPoint(float hue) {
		final RectF rect = hueRect;
		final float height = rect.height();

		Point p = new Point();
		p.y = (int) (height - (hue * height / 360f) + rect.top);
		p.x = (int) rect.left;

		return p;
	}

	private float[] pointToSatVal(float x, float y) {
		float width = satValRect.width();
		float height = satValRect.height();

		x = clamp(x - satValRect.left, 0, width);
		y = clamp(y - satValRect.top, 0, height);

		float[] result = new float[2];
		result[0] = 1.f / width * x;
		result[1] = 1.f - (1.f / height * y);

		return result;
	}

	private float pointToHue(float y) {
		final RectF rect = hueRect;
		float height = rect.height();

		y = clamp(y - rect.top, 0, height);
		return 360f - (y * 360f / height);
	}

	private int[] buildHueColorArray() {
		int[] hue = new int[361];

		int count = 0;
		for (int i = hue.length - 1; i >= 0; i--, count++) {
			hue[count] = Color.HSVToColor(new float[]{i, 1f, 1f});
		}
		return hue;
	}

	private void setUpSatValRect() {
		final RectF dRect = drawingRect;
		float panelContentHeight = dRect.height() - 2 * BORDER_WIDTH_PX;
		float panelContentWidth = dRect.width() - 2 * BORDER_WIDTH_PX - panelSpacing - huePanelWidth;

		float left = dRect.left + BORDER_WIDTH_PX;
		float top = dRect.top + BORDER_WIDTH_PX;
		float bottom = top + panelContentHeight;
		float right = left + panelContentWidth;
		satValRect.set(left, top, right, bottom);
	}

	private void setUpHueRect() {
		final RectF dRect = drawingRect;
		float left = dRect.right - huePanelWidth + BORDER_WIDTH_PX;
		float top = dRect.top + BORDER_WIDTH_PX;
		float bottom = dRect.bottom - BORDER_WIDTH_PX;
		float right = dRect.right - BORDER_WIDTH_PX;

		hueRect.set(left, top, right, bottom);
	}

	public void setOnColorChangedListener(OnColorChangedListener listener) {
		this.onColorChangedListener = listener;
	}

	private void onColorChanged() {
		if (onColorChangedListener != null) {
			onColorChangedListener.colorChanged(getSelectedColor());
		}
	}

	public int getSelectedColor() {
		float[] hsv = {hue, sat, val};
		return Color.HSVToColor(alpha, hsv);
	}

	public void setSelectedColor(int color) {
		float[] hsv = new float[3];
		Color.colorToHSV(color, hsv);
		alpha = Color.alpha(color);
		hue = hsv[0];
		sat = hsv[1];
		val = hsv[2];

		invalidate();
	}

	public interface OnColorChangedListener {
		void colorChanged(int color);
	}
}

File: colorpicker/src/main/java/org/catrobat/paintroid/colorpicker/HSVColorPickerView.kt
/*
 * Paintroid: An image manipulation application for Android.
 * Copyright (C) 2010-2021 The Catrobat Team
 * (<http://developer.catrobat.org/credits>)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.catrobat.paintroid.colorpicker

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.LinearGradient
import android.graphics.Paint
import android.graphics.Point
import android.graphics.PorterDuff
import android.graphics.PorterDuffXfermode
import android.graphics.RectF
import android.graphics.Shader
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View
import kotlin.math.max
import kotlin.math.min

private const val BORDER_WIDTH_PX = 1f
private const val HUE_PANEL_WIDTH = 30f
private const val PANEL_SPACING = 10f
private const val PALETTE_CIRCLE_TRACKING_RADIUS = 5f
private const val RECTANGLE_TRACKER_OFFSET = 2f
private const val HUE = 360f
private const val CHANNELS = 3
private const val ALPHA_MAX = 0xff

class HSVColorPickerView : View {
    private var panelSpacing = PANEL_SPACING
    private var paletteCircleTrackerRadius = PALETTE_CIRCLE_TRACKING_RADIUS
    private var rectangleTrackerOffset = RECTANGLE_TRACKER_OFFSET
    private var huePanelWidth = HUE_PANEL_WIDTH
    private var density = 1f
    private var alpha = ALPHA_MAX
    private var hue = HUE
    private var sat = 0f
    private var brightness = 0f
    private val sliderTrackerColor = -0xa0a0b
    private val borderColor = -0x919192
    private var drawingOffset = 0f
    private var satValPaint: Paint
    private var satValTrackerPaint: Paint
    private var huePaint: Paint
    private var hueTrackerPaint: Paint
    private var borderPaint: Paint
    private var drawingRect: RectF
    private val satValRect: RectF
    private val hueRect: RectF
    private var onColorChangedListener: OnColorChangedListener? = null
    private var onColorFinallySelectedListener: OnColorFinallySelectedListener? = null
    private var startTouchPoint: Point? = null

    var selectedColor: Int
        get() {
            val hsv = floatArrayOf(hue, sat, brightness)
            return Color.HSVToColor(alpha, hsv)
        }
        set(color) {
            val hsv = FloatArray(CHANNELS)
            Color.colorToHSV(color, hsv)
            alpha = Color.alpha(color)
            hue = hsv[0]
            sat = hsv[1]
            brightness = hsv[2]
            invalidate()
        }

    constructor(context: Context?) : super(context)

    constructor(context: Context?, attrs: AttributeSet?) : super(context, attrs)

    constructor(context: Context?, attrs: AttributeSet?, defStyle: Int) : super(
        context,
        attrs,
        defStyle
    )

    init {
        density = context.resources.displayMetrics.density
        paletteCircleTrackerRadius *= density
        rectangleTrackerOffset *= density
        huePanelWidth *= density
        panelSpacing *= density
        drawingOffset = paletteCircleTrackerRadius
        satValTrackerPaint = Paint().apply {
            style = Paint.Style.STROKE
            strokeWidth = 2f * density
            isAntiAlias = true
        }
        hueTrackerPaint = Paint().apply {
            color = sliderTrackerColor
            style = Paint.Style.STROKE
            strokeWidth = 2f * density
            isAntiAlias = true
        }
        satValPaint = Paint()
        huePaint = Paint()
        borderPaint = Paint()
        drawingRect = RectF()
        satValRect = RectF()
        hueRect = RectF()
    }

    private fun clamp(brightness: Float, mn: Float, mx: Float): Float = max(mn, min(mx, brightness))

    private fun drawSatValPanel(canvas: Canvas) {
        borderPaint.color = borderColor
        canvas.drawRect(
            drawingRect.left,
            drawingRect.top,
            satValRect.right + BORDER_WIDTH_PX,
            satValRect.bottom + BORDER_WIDTH_PX,
            borderPaint
        )
        val valShader = LinearGradient(
            satValRect.left,
            satValRect.top,
            satValRect.left,
            satValRect.bottom,
            -0x1,
            -0x1000000,
            Shader.TileMode.CLAMP
        )
        val rgb = Color.HSVToColor(floatArrayOf(hue, 1f, 1f))
        val satShader = LinearGradient(
            satValRect.left,
            satValRect.top,
            satValRect.right,
            satValRect.bottom,
            Color.WHITE,
            rgb,
            Shader.TileMode.CLAMP
        )
        satValPaint.apply {
            xfermode = null
            shader = valShader
            canvas.drawRect(satValRect, this)
            shader = satShader
            xfermode = PorterDuffXfermode(PorterDuff.Mode.MULTIPLY)
            canvas.drawRect(satValRect, this)
        }
        satValToPoint(sat, brightness).apply {
            satValTrackerPaint.color = -0x1000000
            canvas.drawCircle(
                x.toFloat(),
                y.toFloat(),
                paletteCircleTrackerRadius - 1f * density,
                satValTrackerPaint
            )
            satValTrackerPaint.color = -0x222223
            canvas.drawCircle(
                x.toFloat(),
                y.toFloat(),
                paletteCircleTrackerRadius,
                satValTrackerPaint
            )
        }
    }

    private fun drawHuePanel(canvas: Canvas) {
        borderPaint.color = borderColor
        canvas.drawRect(
            hueRect.left - BORDER_WIDTH_PX,
            hueRect.top -
                BORDER_WIDTH_PX,
            hueRect.right + BORDER_WIDTH_PX,
            hueRect.bottom + BORDER_WIDTH_PX,
            borderPaint
        )
        val hueShader = LinearGradient(
            hueRect.left,
            hueRect.top,
            hueRect.left,
            hueRect.bottom,
            buildHueColorArray(),
            null,
            Shader.TileMode.CLAMP
        )
        huePaint.shader = hueShader
        canvas.drawRect(hueRect, huePaint)
        val rectHeight = 2 * density
        hueToPoint(hue).let {
            val r = RectF().apply {
                left = hueRect.left - rectangleTrackerOffset
                right = hueRect.right + rectangleTrackerOffset
                top = it.y - rectHeight
                bottom = it.y + rectHeight
            }
            canvas.drawRoundRect(r, 2f, 2f, hueTrackerPaint)
        }
    }

    private fun moveTrackersIfNeeded(event: MotionEvent): Boolean {
        val startX = startTouchPoint?.x ?: return false
        val startY = startTouchPoint?.y ?: return false
        when {
            hueRect.contains(startX.toFloat(), startY.toFloat()) -> hue = pointToHue(event.y)
            satValRect.contains(startX.toFloat(), startY.toFloat()) -> {
                val result = pointToSatVal(event.x, event.y)
                sat = result[0]
                brightness = result[1]
            }
            else -> return false
        }
        return true
    }

    private fun satValToPoint(sat: Float, brightness: Float): Point {
        val height = satValRect.height()
        val width = satValRect.width()
        return Point().apply {
            x = (sat * width + satValRect.left).toInt()
            y = ((1f - brightness) * height + satValRect.top).toInt()
        }
    }

    private fun hueToPoint(hue: Float): Point {
        val height = hueRect.height()
        return Point().apply {
            x = hueRect.left.toInt()
            y = (height - hue * height / HUE + hueRect.top).toInt()
        }
    }

    private fun pointToSatVal(x: Float, y: Float): FloatArray {
        val width = satValRect.width()
        val height = satValRect.height()
        val curX = clamp(x - satValRect.left, 0f, width)
        val curY = clamp(y - satValRect.top, 0f, height)
        return floatArrayOf(1f / width * curX, 1f - 1f / height * curY)
    }

    private fun pointToHue(y: Float): Float {
        val height = hueRect.height()
        val curY = clamp(y - hueRect.top, 0f, height)
        return HUE - curY * HUE / height
    }

    private fun buildHueColorArray(): IntArray {
        val hue = IntArray(HUE.toInt() + 1)
        var count = 0
        for (i in hue.size - 1 downTo 0) {
            hue[count++] = Color.HSVToColor(floatArrayOf(i.toFloat(), 1f, 1f))
        }
        return hue
    }

    private fun setUpSatValRect() {
        val panelContentHeight =
            drawingRect.height() - 2 * BORDER_WIDTH_PX
        val panelContentWidth =
            drawingRect.width() - 2 * BORDER_WIDTH_PX - panelSpacing - huePanelWidth
        val left = drawingRect.left + BORDER_WIDTH_PX
        val top = drawingRect.top + BORDER_WIDTH_PX
        val bottom = top + panelContentHeight
        val right = left + panelContentWidth
        satValRect.set(left, top, right, bottom)
    }

    private fun setUpHueRect() {
        val left = drawingRect.right - huePanelWidth + BORDER_WIDTH_PX
        val top = drawingRect.top + BORDER_WIDTH_PX
        val bottom = drawingRect.bottom - BORDER_WIDTH_PX
        val right = drawingRect.right - BORDER_WIDTH_PX
        hueRect.set(left, top, right, bottom)
    }

    private fun onColorChanged() {
        onColorChangedListener?.colorChanged(selectedColor)
    }

    fun setOnColorChangedListener(listener: OnColorChangedListener?) {
        onColorChangedListener = listener
    }

    private fun onColorFinallySelected() {
        onColorFinallySelectedListener?.colorFinallySelected(selectedColor)
    }

    fun setOnColorFinallySelectedListener(listener: OnColorFinallySelectedListener?) {
        onColorFinallySelectedListener = listener
    }

    override fun onDraw(canvas: Canvas) {
        if (drawingRect.width() <= 0 || drawingRect.height() <= 0) {
            return
        }
        drawSatValPanel(canvas)
        drawHuePanel(canvas)
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        var update = false
        when (event.action) {
            MotionEvent.ACTION_DOWN -> {
                startTouchPoint = Point(event.x.toInt(), event.y.toInt())
                update = moveTrackersIfNeeded(event)
            }
            MotionEvent.ACTION_MOVE -> update = moveTrackersIfNeeded(event)
            MotionEvent.ACTION_UP -> {
                startTouchPoint = null
                update = moveTrackersIfNeeded(event)
                onColorFinallySelected()
            }
        }
        if (update) {
            invalidate()
            onColorChanged()
            return true
        }
        return super.onTouchEvent(event)
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        drawingRect = RectF().apply {
            left = drawingOffset + paddingLeft
            right = w - drawingOffset - paddingRight
            top = drawingOffset + paddingTop
            bottom = h - drawingOffset - paddingBottom
        }
        setUpSatValRect()
        setUpHueRect()
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        val size = MeasureSpec.getSize(widthMeasureSpec) - paddingStart - paddingEnd
        setMeasuredDimension(size, size)
    }

    fun interface OnColorChangedListener {
        fun colorChanged(color: Int)
    }

    fun interface OnColorFinallySelectedListener {
        fun colorFinallySelected(color: Int)
    }
}
